uvloop/sslproto.pyx

```diff
@@ -471,62 +297,39 @@ class SSLProtocol(object):
         meaning a regular EOF is received or the connection was
         aborted or closed).
         """
-        if self._session_established:
-            self._session_established = False
+        if self._state == _WRAPPED:
             self._loop.call_soon(self._app_protocol.connection_lost, exc)
-
+        else:
+            # Most likely an exception occurred while in SSL handshake.
+            # Just mark the app transport as closed so that its __del__
+            # doesn't complain.
+            if self._app_transport is not None:
+                self._app_transport._closed = True
+        self._set_state(_UNWRAPPED)
         self._transport = None
         self._app_transport = None
         self._wakeup_waiter(exc)
 
-    def pause_writing(self):
-        """Called when the low-level transport's buffer goes over
-        the high-water mark.
-        """
-        self._app_protocol.pause_writing()
+    def get_buffer(self, n):
+        if len(self._ssl_buffer) < n:
+            self._ssl_buffer.extend(
+                0 for _ in range(n - len(self._ssl_buffer)))
+        return self._ssl_buffer
 
-    def resume_writing(self):
-        """Called when the low-level transport's buffer drains below
-        the low-water mark.
-        """
-        self._app_protocol.resume_writing()
+    def buffer_updated(self, nbytes):
+        incoming = memoryview(self._ssl_buffer)[:nbytes]
 
-    def data_received(self, data):
-        """Called when some SSL data is received.
+        self._incoming.write(incoming)
 
-        The argument is a bytes object.
-        """
-        if self._sslpipe is None:
-            # transport closing, sslpipe is destroyed
-            return
+        if self._state == _DO_HANDSHAKE:
+            self._do_handshake()
 
-        try:
-            ssldata, appdata = (<_SSLPipe>self._sslpipe).feed_ssldata(data)
-        except ssl_SSLError as e:
-            msg = (
-                f'SSL error errno:{getattr(e, "errno", "missing")} '
-                f'reason: {getattr(e, "reason", "missing")}'
-            )
-            self._fatal_error(e, msg)
-            return
+        elif self._state == _WRAPPED:
+            self._do_read()
 
-        self._transport.writelines(ssldata)
-
-        for chunk in appdata:
-            if chunk:
-                try:
-                    if self._app_protocol_is_buffer:
-                        _feed_data_to_bufferred_proto(
-                            self._app_protocol, chunk)
-                    else:
-                        self._app_protocol.data_received(chunk)
-                except Exception as ex:
-                    self._fatal_error(
-                        ex, 'application protocol failed to receive SSL data')
-                    return
-            else:
-                self._start_shutdown()
-                break
+        elif self._state == _SHUTDOWN:
+            self._do_read()
+            self._do_shutdown()
```

Comment:
`_do_read()` should be removed in `_SHUTDOWN`, because a peer starts to discard incoming app data once it sends the `close_notify` alert.